### PR TITLE
Untangle static/dynamic state validity tracking a bit

### PR DESCRIPTION
**Pull request by chrisforbes**
This doesn't actually solve #614, but it gets us closer to sanity.

Things still to do here:
- Untangle checks from side effects in CmdBindPipeline
- Untangle compute vs graphics cases (they're not really the same at all)
- Implement the additional check for #614.

Need to consider what to do with the viewport & scissor state for #614 -- we currently handle them differently in order to get per-viewport and per-scissor tracking working properly.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -5130,10 +5130,18 @@ VKAPI_ATTR void VKAPI_CALL CmdBindPipeline(VkCommandBuffer commandBuffer, VkPipe
...
         // TODO: VALIDATION_ERROR_18000612 VALIDATION_ERROR_18000616
 
-        PIPELINE_STATE *pipe_state = getPipelineState(dev_data, pipeline);
+        auto pipe_state = getPipelineState(dev_data, pipeline);
+        if (VK_PIPELINE_BIND_POINT_GRAPHICS == pipelineBindPoint) {
+            auto old_pipe_state = cb_state->lastBound[pipelineBindPoint].pipeline_state;
+            if (old_pipe_state) {
+                cb_state->status &= ~MakeStaticStateMask(old_pipe_state->graphicsPipelineCI.ptr()->pDynamicState);
+            }
+            if (pipe_state) {
```
Comment (tobine) [inline layers/core_validation.cpp]: This "if" check shouldn't be needed as bad pipeline would have triggered in PV. Similar check below is from long, long ago when CV was much more conservative and that check should be able to be killed in favor of same check in PV.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -5132,12 +5133,10 @@ VKAPI_ATTR void VKAPI_CALL CmdBindPipeline(VkCommandBuffer commandBuffer, VkPipe
 
         auto pipe_state = getPipelineState(dev_data, pipeline);
         if (VK_PIPELINE_BIND_POINT_GRAPHICS == pipelineBindPoint) {
-            auto old_pipe_state = cb_state->lastBound[pipelineBindPoint].pipeline_state;
-            if (old_pipe_state) {
-                cb_state->status &= ~MakeStaticStateMask(old_pipe_state->graphicsPipelineCI.ptr()->pDynamicState);
-            }
+            cb_state->status &= ~cb_state->static_status;
```
Comment (tobine) [inline layers/core_validation.cpp]: Yeah, this is good. Extra mask saves the function call.

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -5201,19 +5201,20 @@ VKAPI_ATTR void VKAPI_CALL CmdSetLineWidth(VkCommandBuffer commandBuffer, float
     if (pCB) {
         skip |= ValidateCmdQueueFlags(dev_data, pCB, "vkCmdSetLineWidth()", VK_QUEUE_GRAPHICS_BIT, VALIDATION_ERROR_1d602415);
         skip |= ValidateCmd(dev_data, pCB, CMD_SETLINEWIDTHSTATE, "vkCmdSetLineWidth()");
-        pCB->status |= CBSTATUS_LINE_WIDTH_SET;
 
-        PIPELINE_STATE *pPipeTrav = pCB->lastBound[VK_PIPELINE_BIND_POINT_GRAPHICS].pipeline_state;
-        if (pPipeTrav != NULL && !isDynamic(pPipeTrav, VK_DYNAMIC_STATE_LINE_WIDTH)) {
-            skip |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_WARNING_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT,
+        if (pCB->static_status & CBSTATUS_LINE_WIDTH_SET) {
```
Comment (tobine) [inline layers/core_validation.cpp]: Looks like most of the other CmdSet*() dynamic state functions are missing the check for similar VU. Please follow-on with those check added or open an issue to track them to completion.

Comment (chrisforbes) [inline layers/core_validation.cpp]: I'm adding all the missing cases now.

**Pull request merged**